graphql_compiler/macros/macro_edge/validation.py

```diff
@@ -0,0 +1,249 @@
+    get_schema_for_macro_edge_definitions
+)
+
+
+def _validate_macro_ast_with_macro_directives(schema, ast, macro_directives):
+    """Raise errors if the macro uses the macro directives incorrectly or is otherwise invalid."""
+    if ast.directives:
+        directive_names = [directive.name.value for directive in ast.directives]
+        raise GraphQLInvalidMacroError(
+            u'Unexpectedly found directives at the top level of the GraphQL input. '
+            u'This is not supported. Directives: {}'.format(directive_names))
+
+    if ast.variable_definitions is not None:
+        raise GraphQLInvalidMacroError(
+            u'Unexpectedly found variable definitions at the top level of the GraphQL input. '
+            u'This is not supported. Variable definitions: {}'.format(ast.variable_definitions))
+
+    schema_with_macro_edge_directives = get_schema_for_macro_edge_definitions(schema)
+
+    validation_errors = validate(schema_with_macro_edge_directives, ast)
+    if validation_errors:
+        raise GraphQLInvalidMacroError(
+            u'Macro edge failed validation: {}'.format(validation_errors))
+
+    for directive_definition in DIRECTIVES_REQUIRED_IN_MACRO_EDGE_DEFINITION:
+        macro_data = macro_directives.get(directive_definition.name, None)
+        if not macro_data:
+            raise GraphQLInvalidMacroError(
+                u'Required macro edge directive "@{}" was not found anywhere within the supplied '
+                u'macro edge definition GraphQL.'.format(directive_definition.name))
+
+        if len(macro_data) > 1:
+            raise GraphQLInvalidMacroError(
+                u'Required macro edge directive "@{}" was unexpectedly present more than once in '
+                u'the supplied macro edge definition GraphQL. It was found {} times.'
+                .format(directive_definition.name, len(macro_data)))
+
+
+def _validate_macro_ast_directives(ast, inside_fold_scope=False):
+    """Check that the macro is using non-macro directives properly.
+
+    Restrictions on use of directives:
+    - @output and @output_source are disallowed
+    - @macro_edge_target is not allowed to be inside a @fold scope
+    - @macro_edge_target is not allowed to begin with a coercion
+
+    Args:
+        ast: GraphQL AST describing a subtree of the macro
+        inside_fold_scope: bool, whether the subtree is within a @fold scope
+    """
+    subselection_inside_fold_scope = inside_fold_scope
+    names_of_allowed_directives = frozenset({
+        directive.name
+        for directive in DIRECTIVES_ALLOWED_IN_MACRO_EDGE_DEFINITION
+    })
+
+    for directive in ast.directives:
+        name = directive.name.value
+        if name not in names_of_allowed_directives:
+            raise GraphQLInvalidMacroError(u'Unexpected directive name found: {} {}'
+                                           .format(name, directive))
+
+        if name == FoldDirective.name:
```

Comment:
I think the last example in the `@fold` section of the README is a use case for `@fold` in a macro edge: https://github.com/kensho-technologies/graphql-compiler#fold

Folds don't always have to output data, and the way we'd use them in macro edges would be in that non-outputting capacity since the `@macro_edge_target` can't be inside a scope marked `@fold`.

Reply:
Oh ok makes sense.